book/src/dev/rfcs/drafts/xxxx-async-rust.md

```diff
@@ -0,0 +1,438 @@
+- Feature Name: async_rust_constraints
+- Start Date: 2021-03-30
+- Design PR: [ZcashFoundation/zebra#1965](https://github.com/ZcashFoundation/zebra/pull/1965)
+- Zebra Issue: [ZcashFoundation/zebra#1593](https://github.com/ZcashFoundation/zebra/issues/1593)
+
+# Summary
+[summary]: #summary
+
+Zebra programmers need to carefully write async code so it doesn't deadlock or hang.
+This is particularly important for `poll`, `select`, `Buffer`, `Batch`, and `Mutex`.
+
+Zebra executes concurrent tasks using [async Rust](https://rust-lang.github.io/async-book/),
+with the [tokio](https://docs.rs/tokio/0.3.6/tokio/index.html) executor.
+
+At a higher level, Zebra also uses [`tower::Service`s](https://docs.rs/tower/0.4.1/tower/trait.Service.html),
+[`tower::Buffer`s](https://docs.rs/tower/0.4.1/tower/buffer/struct.Buffer.html),
+and our own [`tower-batch`](https://github.com/ZcashFoundation/zebra/tree/main/tower-batch)
+implementation.
+
+# Motivation
+[motivation]: #motivation
+
+Like all concurrent codebases, Zebra needs to obey certain constraints to avoid
+hangs. Unfortunately, Rust's tooling in these areas is still developing. So
+Zebra developers need to manually check these constraints during design,
+development, reviews, and testing.
+
+# Definitions
+[definitions]: #definitions
+
+- `hang`: a Zebra component stops making progress.
+- `constraint`: a rule that Zebra must follow to prevent `hang`s.
+- `CORRECTNESS comment`: the documentation for a `constraint` in Zebra's code.
+- `task`: an async task can execute code independently of other tasks, using
+        cooperative multitasking.
+- `deadlock`: a `hang` that stops an async task executing code.
+        For example: a task is never woken up.
+- `starvation` or `livelock`: a `hang` that executes code, but doesn't do
+        anything useful. For example: a loop never terminates.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you are designing, developing, or testing concurrent Zebra code, follow the
+patterns in these examples to avoid hangs.
+
+If you are reviewing concurrent Zebra designs or code, make sure that:
+- it is clear how the design or code avoids hangs
+- the design or code follows the patterns in these examples (as much as possible)
+- the concurrency constraints and risks are documented
+
+The [Reference](#reference-level-explanation) section contains in-depth
+background information about Rust async concurrency in Zebra.
+
+Here are some examples of concurrent designs and documentation in Zebra:
+
+## Registering Wakeups Before Returning Poll::Pending
+[wakeups-poll-pending]: #wakeups-poll-pending
+
+Here's a wakeup correctness example from
+[unready_service.rs](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/unready_service.rs#L43):
```

Comment:
I think we should not link to `main` but to the release: https://github.com/ZcashFoundation/zebra/blob/v1.0.0-alpha.5/zebra-network/src/peer_set/unready_service.rs as a general rule in RFCs.

However, in this case, the code is not yet present in the last release so if we merge before the release we should add a follow up ticket to change it after the tag is done.

Reply:
Thanks for picking this up, I'll switch from using `main` to using the commit hash.
```suggestion
[unready_service.rs](https://github.com/ZcashFoundation/zebra/blob/de6d1c93f3e4f9f4fd849176bea6b39ffc5b260f/zebra-network/src/peer_set/unready_service.rs#L43):
```

Reply:
Commit hash will also work. Thanks.